### PR TITLE
Roster query builder

### PR DESCRIPTION
**Pull request by AnthonyWhitaker**
This is one of those tasks I could spend another sprint or two on... committing now due to sprint end.

List, String, Custom Columns, Boolean, Number and Date Value editors
image

Column Selection
image

### CODE REVIEW

File: src/hooks/use-deep-equality.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import {
+  useState, useEffect, SetStateAction, Dispatch,
+} from 'react';
+import deepEquals from 'fast-deep-equal';
+
+export default function useDeepEquality<T>(value?: T | undefined): [T| undefined, Dispatch < SetStateAction < T| undefined >>, T| undefined] {
```
Comment (larskendall) [inline src/hooks/use-deep-equality.ts:6]: The spacing seems a little odd in the return type here. Does this work ok instead?
```typescript
export default function useDeepEquality<T>(value?: T | undefined): [T | undefined, Dispatch<SetStateAction<T | undefined>>, T | undefined] {

```

**Review by larskendall — CHANGES_REQUESTED** [review]
The functionality is all looking great. There were just a few minor linting type things I noticed.

**Comment by DaveJackson-PE** [comment]
After seeing a number of lint issues in these reviews, I realized that I needed to install **AND** _**enable**_ the eslint extension in VS Code. Not sure if this is an issue for @AnthonyWhitaker or anyone else on the team, but thought I'd share just in case...

**Comment by larskendall** [comment]
@DaveJackson-PE Yeah it's best to have eslint flag stuff right in your IDE and/or run eslint manually to check for warnings before submitting a PR. There are still a couple wonky project setup issues that might make it a little awkward to run manually, so it's probably easiest to just rely on the IDE for now until we can make another technical debt pass. Ideally we should also automatically flag them all on GitHub when submitting a PR.

**Comment by AnthonyWhitaker** [comment]
> After seeing a number of lint issues in these reviews, I realized that I needed to install **AND** _**enable**_ the eslint extension in VS Code. Not sure if this is an issue for @AnthonyWhitaker or anyone else on the team, but thought I'd share just in case...

I do have or enabled. There's something weird about the way vscode or the eslint plugin seems to be handling the nested config for the server. If usually formats in sace and so should have caught the double quotes and any other auto-fixable items. I took a note to look into it but haven't made time.

File: server/api/roster/roster.controller.ts [hunk 2]
```diff
@@ -180,6 +196,94 @@ class RosterController {
     });
   }
 
+  async searchRoster(req: ApiRequest<OrgParam, SearchRosterBody, GetRosterQuery>, res: Response) {
+    const limit = (req.query.limit != null) ? parseInt(req.query.limit) : 100;
+    const page = (req.query.page != null) ? parseInt(req.query.page) : 0;
```
[the remark this answers is not in the document]
Comment (AnthonyWhitaker) [reply server/api/roster/roster.controller.ts:201]: Going to delay this change for now. This code is actually just copy-pasted from getRoster() on line 176 - which is the non-filtered equivalent of this call. I think in the future these should be consolidated or refactored in some other way.

File: server/api/roster/roster.controller.ts [hunk 3]
```diff
@@ -18,6 +18,22 @@ import { Role } from '../role/role.model';
 import { Unit } from '../unit/unit.model';
 
 class RosterController {
+  static getColumnSelect(column: RosterColumnInfo) {
```
[the remark this answers is not in the document]
Comment (AnthonyWhitaker) [reply server/api/roster/roster.controller.ts:21]: I can't get my IDE to work with our eslint/tsconfig setup. src/* seems to work but anything in server/ is failing. I'll take a pass at this now - manually but will get WebStorm installed or work on fixing the vscode setup for future changes.

**Pull request merged**